pkg/istio-agent/agent.go

```diff
@@ -474,26 +476,39 @@ func fileExists(path string) bool {
 }
 
 // Find the root CA to use when connecting to the CA (Istiod or external).
-func (a *Agent) FindRootCAForCA() string {
+func (a *Agent) FindRootCAForCA() (string, error) {
+	var rootCAPath string
+
 	if a.cfg.CARootCerts == security.SystemRootCerts {
-		return ""
+		return "", nil
 	} else if a.cfg.CARootCerts != "" {
-		return a.cfg.CARootCerts
-	} else if a.secOpts.PilotCertProvider == "kubernetes" {
+		rootCAPath = a.cfg.CARootCerts
+	} else if a.secOpts.PilotCertProvider == constants.CertProviderKubernetes {
 		// Using K8S - this is likely incorrect, may work by accident.
-		// API is alpha.
-		return k8sCAPath // ./var/run/secrets/kubernetes.io/serviceaccount/ca.crt
-	} else if a.secOpts.PilotCertProvider == "custom" {
-		return security.DefaultRootCertFilePath // ./etc/certs/root-cert.pem
+		// API is GA.
+		rootCAPath = k8sCAPath // ./var/run/secrets/kubernetes.io/serviceaccount/ca.crt
+	} else if a.secOpts.PilotCertProvider == constants.CertProviderCustom {
+		rootCAPath = security.DefaultRootCertFilePath // ./etc/certs/root-cert.pem
 	} else if a.secOpts.ProvCert != "" {
 		// This was never completely correct - PROV_CERT are only intended for auth with CA_ADDR,
 		// and should not be involved in determining the root CA.
-		return a.secOpts.ProvCert + "/root-cert.pem"
+		// For VMs, the root cert file used to auth may be populated afterwards.
+		// Thus, return directly here and skip checking for existence.
+		return a.secOpts.ProvCert + "/root-cert.pem", nil
+	} else if a.secOpts.PilotCertProvider == constants.CertProviderNone {
+		return "", fmt.Errorf("root CA file for CA required but configured provider as none")
 	} else {
 		// This is the default - a mounted config map on K8S
-		return path.Join(CitadelCACertPath, constants.CACertNamespaceConfigMapDataName)
+		rootCAPath = path.Join(CitadelCACertPath, constants.CACertNamespaceConfigMapDataName)
```

Comment:
The original code does not check the existence of the root cert file for the mounted config map (the original code just returns the path). The PR checks the existence of the root cert file for the mounted config map. Quoted from the comment "// For VMs, the root cert file used to auth may be populated afterwards", similar to the root cert file for VM, is it possible that the root cert file for the mounted config map may be populated afterwards?

Reply:
IMO, I think this should be available as part of the startup. Any comment? @howardjohn @ramaraochavali

Reply:
I am not aware of how this config map works.